src/dataframerow/dataframerow.jl

```diff
@@ -37,49 +37,76 @@ Base.done(r::DataFrameRow, s) = s > length(r)
+    h = hash_colel(cols[1], r, h)
+    rowhash(Base.tail(cols), r, h)
 end
 
-# compare two elements in the pooled array
-# NOTE assume there are no duplicated elements in the pool
-_isequalelms(a::PooledDataArray, i::Int, j::Int) = isequal(a.refs[i], a.refs[j])
+Base.hash(r::DataFrameRow, h::UInt = zero(UInt)) =
+    rowhash(ntuple(i -> r.df[i], ncol(r.df)), r.row, h)
 
 # comparison of DataFrame rows
 # only the rows of the same DataFrame could be compared
 # rows are equal if they have the same values (while the row indices could differ)
+# if all non-null values are equal, but there are nulls, returns null
+Base.:(==)(r1::DataFrameRow, r2::DataFrameRow) = isequal(r1, r2)
+
 function Base.isequal(r1::DataFrameRow, r2::DataFrameRow)
-    if r1.df !== r2.df
-        throw(ArgumentError("Comparing rows from different frames not supported"))
+    isequal_row(r1.df, r1.row, r2.df, r2.row)
+end
+
+# internal method for comparing the elements of the same data table column
+isequal_colel(col::AbstractArray, r1::Int, r2::Int) =
+    (r1 == r2) || isequal(Base.unsafe_getindex(col, r1), Base.unsafe_getindex(col, r2))
+
+# table columns are passed as a tuple of vectors to ensure type specialization
+isequal_row(cols::Tuple{AbstractVector}, r1::Int, r2::Int) =
+    isequal(cols[1][r1], cols[1][r2])
+isequal_row(cols::Tuple{Vararg{AbstractVector}}, r1::Int, r2::Int) =
+    isequal(cols[1][r1], cols[1][r2]) && isequal_row(Base.tail(cols), r1, r2)
+
+isequal_row(cols1::Tuple{AbstractVector}, r1::Int, cols2::Tuple{AbstractVector}, r2::Int) =
+    isequal(cols1[1][r1], cols2[1][r2])
+isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+            cols2::Tuple{Vararg{AbstractVector}}, r2::Int) =
+    isequal(cols1[1][r1], cols2[1][r2]) &&
+        isequal_row(Base.tail(cols1), r1, Base.tail(cols2), r2)
+
+function isequal_row(df1::AbstractDataFrame, r1::Int, df2::AbstractDataFrame, r2::Int)
+    if df1 === df2
+        if r1 == r2
+            return true
+        end
+    elseif !(ncol(df1) == ncol(df2))
+        throw(ArgumentError("Rows of the tables that have different number of columns cannot be compared. Got $(ncol(df1)) and $(ncol(df2)) columns"))
     end
-    if r1.row == r2.row
-        return true
+    @inbounds for (col1, col2) in zip(columns(df1), columns(df2))
+        isequal(col1[r1], col2[r2]) || return false
     end
-    for col in columns(r1.df)
-        if !_isequalelms(col, r1.row, r2.row)
-            return false
-        end
+    return true
+end
+
+# lexicographic ordering on DataFrame rows, null > !null
+function Base.isless(r1::DataFrameRow, r2::DataFrameRow)
+    (ncol(r1.df) == ncol(r2.df)) ||
+        throw(ArgumentError("Rows of the data tables that have different number of columns cannot be compared ($(ncol(df1)) and $(ncol(df2)))"))
+    @inbounds for i in 1:ncol(r1.df)
+        isless(r1.df[i][r1.row], r2.df[i][r2.row]) && return true
```

Comment:
Since we're doing a general review, we should have a look at the semantics of this function at some point. It was added in https://github.com/JuliaData/DataTables.jl/pull/17 and I'm not sure what's its purpose. Its behavior looks quite counter-intuitive to me.

For now I've kept it, but simplified it a bit since `null` uses the right `isless` definition.

Reply:
See https://github.com/JuliaStats/DataFrames.jl/pull/1222. Actually, the way the code was written looked weird to me, but the behavior was OK.